feat: Implement the prompt module for jobs

johnletey opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Description
<!--- Describe your changes in detail -->

I have implemented the prompt module that shows a ✦ when you have a job running

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes #80

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Screenshots (if appropriate):

<!-- ![Shot 2019-08-11 at 08 38 19](https://user-images.githubusercontent.com/30328854/62831207-61a34000-bc13-11e9-97e7-2d9deeb636de.jpg) -->

![Shot 2019-08-12 at 17 45 39](https://user-images.githubusercontent.com/30328854/62882106-1418f700-bd29-11e9-9f2e-c74a2c7b23a6.jpg)

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [x] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

matchai commented:
Hey @johnletey. How are you doing with this module? Do you need a hand getting this working?

johnletey commented:
> Hey @johnletey. How are you doing with this module? Do you need a hand getting this working?

Thanks so much @matchai ... yes, as a matter of fact, I would love some help getting this working!

matchai commented:
What exactly do you need help with? What does the PR currently do?

johnletey commented:
@matchai Currently the code doesn't show anything (even when I have it print when there is no job currently running)

matchai commented:
Running `jobs` with a spawned command might only show the jobs that are running in bash. I would recommend passing the ongoing jobs as a flag in `starship init`.

johnletey commented:
@matchai Do you mean `starship init` or `starship prompt`?

matchai commented:
I mean `starship init`. We would pass the output of a function in the user's shell, much like how `$status` is passed in as a flag because it isn't exported by its shell: https://github.com/starship/starship/blob/master/src/init.rs#L30

johnletey commented:
@matchai Did I do it right? I still need to display something for different values of the `--job` parameter, but did I at least instantiate and set everything up correctly?

matchai commented:
Yep! That's what I had in mind there.

johnletey commented:
@matchai Wonderful ... sorry, but ran into some problems running locally on my computer ... is the newly added code correct?

matchai commented on src/modules/jobs.rs:
```diff
@@ -1,35 +1,21 @@
-            module.set_style(module_color);
+    let arguments = &context.arguments;
+    if arguments.value_of("jobs").unwrap_or("0") == "0" {
+        module.new_segment("symbol", "");
```
At this point, I'd probably suggest returning `None`.

johnletey replied:
Sounds good @matchai!

johnletey commented:
Updated @matchai!

matchai commented on src/modules/jobs.rs:
```diff
@@ -0,0 +1,25 @@
+    let num_of_jobs = arguments.value_of("jobs").unwrap_or("0");
+    if num_of_jobs == "0" {
+        return None;
+    } else {
```
I would suggest removing this `else {`.
The code should behave the same but will involve less nesting.

johnletey replied:
So something like this?

```rust
if num_of_jobs != "0" {
  module.new_segment("symbol", PROMPT_CHAR);
  if num_of_jobs != "1" {
    module.new_segment("number", num_of_jobs);
  }
};
```

johnletey replied:
nvm @matchai ... figured it out

matchai commented on src/main.rs:
```diff
@@ -22,6 +22,13 @@ fn main() {
         .help("The status code of the previously run command")
         .takes_value(true);
 
+    let job_arg = Arg::with_name("job")
```
I would consider naming this flag `jobs`, as we're expecting a count of potentially multiple jobs.

johnletey replied:
will do

johnletey replied:
done @matchai!

matchai commented on src/main.rs:
```diff
@@ -22,6 +22,13 @@ fn main() {
+        .long("job")
+        .value_name("JOB")
+        .help("The number of currently running jobs")
+        .takes_value(0);
```
Why the `.takes_value(0)`? 🤔 
afaik, `takes_value()` accepts a boolean value.

johnletey replied:
You are right ... setting to `false` by default

matchai commented on src/modules/jobs.rs:
```diff
@@ -0,0 +1,25 @@
+use super::{Context, Module};
+
+/// Creates a segment to show if there are any active jobs running
+pub fn segment<'a>(context: &'a Context) -> Option<Module<'a>> {
```
It looks like unit tests are failing because of this:
```suggestion
pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
```

I would take the time to review our continuous integration.

matchai commented:
This feature is almost at the point of being mergable, but it seems CI is still failing. You can take a look at the results here: https://dev.azure.com/starship-control/starship/_build/results?buildId=1067&view=results

johnletey commented:
@matchai Tests are now passing!

matchai commented:
Excellent!
The last things that remain are documentation and acceptance tests:
- Acceptance tests would be added in a `jobs.rs` file in `tests/testsuite`
- A new heading would be made in `docs/config/README.md`
- `README.md` would be updated to no longer say that this is a planned feature

johnletey commented:
@matchai For the docs, should I update the code so that users can set a custom symbol for jobs?

matchai commented:
Yeah, I think that would be some great configuration to include!
Another possible configuration option is the threshold above which the job count is shown (`1` by default).

johnletey commented:
@matchai I've added tests + configurations to this module!

The tests don't seem to be passing because my module is always returning nothing. I have a suspicion that it has to do with the fact that I'm typecasting quite a lot, but would love your opinion on it!

Also, I have also updated the docs, but I don't like the wording, so would also love to have you review that as well!

johnletey commented:
@matchai All the tests are passing! Would love a review now 😄

matchai commented on src/modules/mod.rs:
```diff
@@ -30,6 +31,7 @@ pub fn handle<'a>(module: &str, context: &'a Context) -> Option<Module<'a>> {
         "username" => username::module(context),
         "battery" => battery::module(context),
         "cmd_duration" => cmd_duration::module(context),
+        "job" | "jobs" => jobs::module(context),
```
We're soon going to be making all modules only render with a single specific name, so let's to that in advance with this module.
```suggestion
        "jobs" => jobs::module(context),
```

matchai commented on src/print.rs:
```diff
@@ -19,6 +19,7 @@ const PROMPT_ORDER: &[&str] = &[
     "python",
     "go",
     "cmd_duration",
+    "jobs",
```
Much like in spaceship, I think this would be better placed after `line_break`, so it would appear beside the character.

johnletey replied:
@matchai I've made this change, but in my opinion, it looks kind of weird:

![Shot 2019-08-12 at 08 47 50](https://user-images.githubusercontent.com/30328854/62851437-fa9e8d80-bcdd-11e9-9fc5-2cb17941e540.jpg)

matchai replied:
Here's what the implementation from spaceship looks like, which I think is a little closer to what I had in mind.

![image](https://user-images.githubusercontent.com/4658208/62881048-d605de00-bcfc-11e9-89e6-bf80ef35183d.png)

Perhaps we can strip away the spacing after the `✦` and the prefix of the module.

johnletey replied:
Understand @matchai ... I like that a lot better ... I have updated the code

matchai commented on tests/testsuite/jobs.rs:
```diff
@@ -0,0 +1,85 @@
+use ansi_term::Color;
```
Tests look great. 🎉

johnletey commented on src/init.rs:
```diff
@@ -108,6 +108,6 @@ function fish_prompt;
     set -l CMD_DURATION "$CMD_DURATION$cmd_duration";
     set -l starship_duration (math --scale=0 "$CMD_DURATION / 1000");
-    starship prompt --status=$exit_code --cmd-duration=$starship_duration;
+    starship prompt --status=$exit_code --cmd-duration=$starship_duration --jobs=$(count (jobs -p));
```
Got it @matchai ... I have no clue how Fish works, so I just pushed your suggestion 😄

matchai replied:
No worries! I would suggest installing fish for testing these init functions in these situations. 👍

johnletey commented on src/main.rs:
```diff
@@ -43,6 +43,13 @@ fn main() {
+    let jobs_arg = Arg::with_name("jobs")
+        .short("j")
+        .long("jobs")
+        .value_name("JOB")
```
Thanks!

johnletey commented on src/modules/jobs.rs:
```diff
@@ -0,0 +1,29 @@
+
+    let arguments = &context.arguments;
+    let num_of_jobs = arguments.value_of("jobs").unwrap_or("0");
+    let num_of_jobs: i64 = num_of_jobs.parse().unwrap();
```
Fixed
